File: src/main/java/com/xyfindables/ui/views/XYSignalBars.java
package com.xyfindables.ui.views;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.RectF;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;

import com.xyfindables.ui.R;

public class XYSignalBars extends View {

    final private static String TAG = XYSignalBars.class.getSimpleName();

    private int _barMaxCount = 3;
    private int _barCount = 0;

    private Paint paintStroke = null;

    private Paint paintEmpty = null;

    private Paint paintFill = null;

    final private static float _barWidthPercent = 0.70f;

    private RectF _rect = new RectF();

    public XYSignalBars(Context context) {
        this(context, null, 0);
    }

    public XYSignalBars(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public XYSignalBars(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        int color = Color.WHITE;
        if (attrs != null) {
            int value = attrs.getAttributeResourceValue("http://schemas.android.com/apk/res/android", "color", 0);
            if (value != 0) {
                color = getResources().getColor(value);
            } else {
                color = attrs.getAttributeIntValue("http://schemas.android.com/apk/res/android", "color", Color.WHITE);
            }
        }

        if (attrs != null) {
            TypedArray attributyeArray = context.obtainStyledAttributes(
                    attrs,
                    R.styleable.XYSignalBars,
                    0, 0);

            if (attributyeArray != null) {
                _barMaxCount = attributyeArray.getInt(R.styleable.XYSignalBars_maxBars, 3);
                attributyeArray.recycle();
            }
        }

        paintStroke = new Paint(Paint.ANTI_ALIAS_FLAG);
        paintStroke.setStyle(Style.STROKE);
        paintStroke.setStrokeWidth(2);

        paintEmpty = new Paint(Paint.ANTI_ALIAS_FLAG);
        paintEmpty.setAlpha(96);
        paintEmpty.setStyle(Style.FILL);

        paintFill = new Paint(Paint.ANTI_ALIAS_FLAG);
        paintFill.setAlpha(192);
        paintFill.setStyle(Style.FILL);

        setColor(color);
    }

    public void setColor(int color) {
        paintStroke.setColor(color);
        paintStroke.setAlpha(255);
        paintEmpty.setColor(color);
        paintEmpty.setAlpha(96);
        paintFill.setColor(color);
        paintFill.setAlpha(192);
    }

    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);

        float width = getMeasuredWidth() - getPaddingLeft() - getPaddingRight();
        float height = getMeasuredHeight() - getPaddingTop() - getPaddingBottom();

        float barWidth = width / _barMaxCount;
        float barStep = height / (_barMaxCount);

        for (int i = 0; i < _barMaxCount; i++) {
            float left = (i * barWidth) + getPaddingLeft();
            float top = getPaddingTop() + (((_barMaxCount - 1) - i) * barStep);

            _rect.left = left + (barWidth * (1 - _barWidthPercent)) / 2;
            _rect.top = top;

            _rect.right = left + barWidth * _barWidthPercent + (barWidth * (1 - _barWidthPercent)) / 2;
            _rect.bottom = getMeasuredHeight() - getPaddingBottom();

            Paint currPaint = i < _barCount ? paintFill : paintEmpty;
            int radius = dpToPx(getResources(), 2);
            canvas.drawRoundRect(_rect, radius, radius, currPaint);
        }
    }

    public void setBarCount(int barCount) {
        if (barCount != _barCount) {
            _barCount = barCount;
            invalidate();
        }
    }

    public int dpToPx(@NonNull Resources res, int dp) {
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, res.getDisplayMetrics());
    }

}

File: ui-android-library/src/main/java/com/xyfindables/ui/views/XYSignalBars.kt
package com.xyfindables.ui.views

import android.content.Context
import android.content.res.Resources
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.Style
import android.graphics.RectF
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.animation.AlphaAnimation
import com.xyfindables.core.XYBase

import com.xyfindables.ui.R

class XYSignalBars @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : View(context, attrs, defStyle) {

    private var barMaxCount = 3
    private var barCount = 0

    private var paintStroke: Paint? = null

    private var paintEmpty: Paint? = null

    private var paintFill: Paint? = null

    private val rect = RectF()

    init {

        var color = Color.WHITE
        if (attrs != null) {
            val value = attrs.getAttributeResourceValue("http://schemas.android.com/apk/res/android", "color", 0)
            if (value != 0) {
                color = @Suppress("DEPRECATION")resources.getColor(value)
            } else {
                color = attrs.getAttributeIntValue("http://schemas.android.com/apk/res/android", "color", Color.WHITE)
            }
        }

        if (attrs != null) {
            val attributyeArray = context.obtainStyledAttributes(
                    attrs,
                    R.styleable.XYSignalBars,
                    0, 0)

            if (attributyeArray != null) {
                barMaxCount = attributyeArray.getInt(R.styleable.XYSignalBars_maxBars, 3)
                attributyeArray.recycle()
            }
        }

        paintStroke = Paint(Paint.ANTI_ALIAS_FLAG)
        paintStroke!!.style = Style.STROKE
        paintStroke!!.strokeWidth = 2f

        paintEmpty = Paint(Paint.ANTI_ALIAS_FLAG)
        paintEmpty!!.alpha = 96
        paintEmpty!!.style = Style.FILL

        paintFill = Paint(Paint.ANTI_ALIAS_FLAG)
        paintFill!!.alpha = 192
        paintFill!!.style = Style.FILL

        setColor(color)
    }

    fun setColor(color: Int) {
        paintStroke!!.color = color
        paintStroke!!.alpha = 255
        paintEmpty!!.color = color
        paintEmpty!!.alpha = 96
        paintFill!!.color = color
        paintFill!!.alpha = 192
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val width = (measuredWidth - paddingLeft - paddingRight).toFloat()
        val height = (measuredHeight - paddingTop - paddingBottom).toFloat()

        val barWidth = width / barMaxCount
        val barStep = height / barMaxCount

        for (i in 0 until barMaxCount) {
            val left = i * barWidth + paddingLeft
            val top = paddingTop + (barMaxCount - 1 - i) * barStep

            rect.left = left + barWidth * (1 - barWidthPercent) / 2
            rect.top = top

            rect.right = left + barWidth * barWidthPercent + barWidth * (1 - barWidthPercent) / 2
            rect.bottom = (measuredHeight - paddingBottom).toFloat()

            val currPaint = if (i < barCount) paintFill else paintEmpty
            val radius = dpToPx(resources, 2)
            canvas.drawRoundRect(rect, radius.toFloat(), radius.toFloat(), currPaint)
        }
    }

    fun setBarCount(updatedBarCount: Int, animate:Boolean) {
        XYBase.logInfo("arie", "setBarCount: $updatedBarCount")
        if (barCount != updatedBarCount) {
            barCount = updatedBarCount
            invalidate()
        }
        if (animate) {
            val animation = AlphaAnimation(0.0f, 1.0f)
            animation.setDuration(500)
            this.startAnimation(animation)
        }
    }

    fun dpToPx(res: Resources, dp: Int): Int {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), res.displayMetrics).toInt()
    }

    companion object {
        private val barWidthPercent = 0.70f
    }

}
